File: cpp/PhotonLibOS/channel.cpp
// Test performance of the channel / async queue primitive.
//
// Canonical (TooManyCooks) implementation: ../TooManyCooks/channel.cpp
//
// N producers each post a slice of kElementCount items; M consumers pull
// until the channel is drained; the total count and sum are validated.
//
// Photon's native MPMC primitive is photon::common::RingChannel, a
// photon-aware wrapper (yield-then-park on empty) around its lock-free MPMC
// ring queue - the same construct Photon's own WorkPool dispatches through.
// Like Go channels it is bounded; we give it the same generous 65536-slot
// buffer as the other bounded-channel implementations so the benchmark
// measures steady-state throughput rather than constant task parking.
// RingChannel has no close/drain notion, so after the producers finish, one
// poison value per consumer is pushed; a consumer drains until it pops one,
// which mirrors draining a closed Go channel.
//
// Producers and consumers are bench::threads, distributed round-robin
// across the work-pool vcpus on creation.

#include "memusage.hpp"
#include "photon_bench.hpp"

#include <photon/common/lockfree_queue.h>

#include <chrono>
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <memory>
#include <vector>

static std::size_t thread_count;
static std::size_t producer_count = 4;
static std::size_t consumer_count = 4;
static const std::size_t iter_count = 1;

static constexpr std::size_t element_count = 10000000;
static constexpr std::size_t channel_buffer = std::size_t{1} << 16;

// Values are 0..element_count-1, so this can't collide with real data.
static constexpr std::uint64_t poison = ~std::uint64_t{0};

using Queue =
  photon::common::RingChannel<LockfreeMPMCRingQueue<std::uint64_t, channel_buffer>>;

struct result {
  std::uint64_t count;
  std::uint64_t sum;
};

static std::uint64_t do_bench() {
  auto queue = std::make_unique<Queue>();

  std::vector<bench::thread> producers;
  producers.reserve(producer_count);
  const std::size_t per_task = element_count / producer_count;
  const std::size_t rem = element_count % producer_count;
  std::uint64_t base = 0;
  for (std::size_t i = 0; i < producer_count; ++i) {
    const std::uint64_t count = i < rem ? per_task + 1 : per_task;
    producers.emplace_back([&queue, count, base] {
      for (std::uint64_t j = 0; j < count; ++j) {
        queue->send<PhotonPause>(base + j);
      }
    });
    base += count;
  }

  std::vector<result> results(consumer_count);
  std::vector<bench::thread> consumers;
  consumers.reserve(consumer_count);
  for (std::size_t i = 0; i < consumer_count; ++i) {
    result* slot = &results[i];
    consumers.emplace_back([&queue, slot] {
      result r{0, 0};
      for (;;) {
        const std::uint64_t value = queue->recv();
        if (value == poison) {
          break;
        }
        ++r.count;
        r.sum += value;
      }
      *slot = r;
    });
  }

  // Join the producers, then poison the queue once per consumer so the
  // consumers' recv() loops terminate after draining.
  for (auto& producer : producers) {
    producer.join();
  }
  for (std::size_t i = 0; i < consumer_count; ++i) {
    queue->send<PhotonPause>(poison);
  }

  std::uint64_t count = 0;
  std::uint64_t sum = 0;
  for (std::size_t i = 0; i < consumer_count; ++i) {
    consumers[i].join();
    count += results[i].count;
    sum += results[i].sum;
  }

  const std::uint64_t expected_sum =
    static_cast<std::uint64_t>(element_count) * (element_count - 1) / 2;
  if (count != element_count) {
    std::printf(
      "FAIL: Expected %zu elements but consumed %" PRIu64 " elements\n",
      element_count, count
    );
  }
  if (sum != expected_sum) {
    std::printf(
      "FAIL: Expected %" PRIu64 " sum but got %" PRIu64 " sum\n", expected_sum,
      sum
    );
  }
  return sum;
}

int main(int argc, char* argv[]) {
  thread_count = bench::default_thread_count();
  if (argc > 1) {
    thread_count = static_cast<std::size_t>(atoi(argv[1]));
  }
  std::size_t c = thread_count / 2;
  if (c == 0) {
    c = 1;
  }
  producer_count = c;
  consumer_count = c;

  std::printf("threads: %zu\n", thread_count);
  std::printf("producers: %zu\n", producer_count);
  std::printf("consumers: %zu\n", consumer_count);

  bench::quiet_logs();
  bench::use_pooled_stacks();

  bench::run_in_pool(thread_count, [] {
    {
      auto result = do_bench(); // warmup
      std::printf("output: %" PRIu64 "\n", result);
    }

    auto startTime = std::chrono::high_resolution_clock::now();
    for (std::size_t i = 0; i < iter_count; ++i) {
      (void)do_bench();
    }
    auto endTime = std::chrono::high_resolution_clock::now();
    auto totalTimeUs = std::chrono::duration_cast<std::chrono::microseconds>(
                         endTime - startTime
    )
                         .count();

    const std::size_t elementsPerSec = static_cast<std::size_t>(
      static_cast<double>(element_count) * 1000000.0 /
      static_cast<double>(totalTimeUs)
    );
    std::printf("runs:\n");
    std::printf("  - iteration_count: %zu\n", iter_count);
    std::printf("    elements: %zu\n", element_count);
    std::printf("    duration: %" PRIu64 " us\n",
                static_cast<std::uint64_t>(totalTimeUs));
    std::printf("    elements/sec: %zu\n", elementsPerSec);
    std::printf("    max_rss: %ld KiB\n", peak_memory_usage());
  });
}

File: rust/beekeeper/src/bin/channel.rs
// Test performance of the library's async queue primitive.
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/channel.cpp
//
// N producers each post a slice of `element_count` items; M consumers pull
// until the queue is drained; the total count and sum are validated.
//
// beekeeper is a worker pool, not a channel library, but its core primitive IS
// an MPMC pipeline: inputs flow through the hive's task queues to the worker
// threads, and each result flows through the outcome channel to whoever
// receives it. This benchmark measures that pipeline end-to-end: producer
// threads submit u64 elements to a hive of `EchoWorker`s (the identity
// worker), and consumer threads drain the hive's outcome channel, summing the
// values. Producers and consumers are external OS threads (submission and
// outcome receipt cannot run on hive workers); the hive's worker threads are
// the benchmark's thread count, as with the other runtimes.

use beekeeper::bee::stock::EchoWorker;
use beekeeper::hive::outcome_channel;
use beekeeper_bench::{
    build_hive, default_threads, join_outcome_drainers, peak_memory_usage, spawn_outcome_drainers,
    WorkstealingHive,
};
use std::time::Instant;

const ELEMENT_COUNT: u64 = 10_000_000;
const ITER_COUNT: usize = 1;

type ChannelWorker = EchoWorker<u64>;

// Split ELEMENT_COUNT into per-producer (count, base) work assignments.
fn producer_assignments(producer_count: usize) -> Vec<(u64, u64)> {
    let per = ELEMENT_COUNT / producer_count as u64;
    let rem = ELEMENT_COUNT % producer_count as u64;
    let mut base = 0u64;
    let mut out = Vec::with_capacity(producer_count);
    for i in 0..producer_count as u64 {
        let count = if i < rem { per + 1 } else { per };
        out.push((count, base));
        base += count;
    }
    out
}

fn do_bench(
    hive: &WorkstealingHive<ChannelWorker>,
    producer_count: usize,
    consumer_count: usize,
) -> u64 {
    let (tx, rx) = outcome_channel::<ChannelWorker>();
    let consumers = spawn_outcome_drainers(&rx, consumer_count);
    drop(rx);

    let producers: Vec<_> = producer_assignments(producer_count)
        .into_iter()
        .map(|(count, base)| {
            // The hive is shallow-cloneable: all clones share the same queues
            // and worker threads, like cloning a channel sender.
            let hive = hive.clone();
            let tx = tx.clone();
            std::thread::spawn(move || {
                for i in 0..count {
                    hive.apply_send(base + i, &tx);
                }
            })
        })
        .collect();
    drop(tx);

    for p in producers {
        p.join().unwrap();
    }
    // Every in-flight task holds a clone of tx, so the consumers finish when
    // the last element's outcome has been delivered.
    let (count, sum) = join_outcome_drainers(consumers);

    let expected_sum = ELEMENT_COUNT * (ELEMENT_COUNT - 1) / 2;
    if count != ELEMENT_COUNT {
        println!(
            "FAIL: Expected {} elements but consumed {} elements",
            ELEMENT_COUNT, count
        );
    }
    if sum != expected_sum {
        println!("FAIL: Expected {} sum but got {} sum", expected_sum, sum);
    }
    sum
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        default_threads()
    };

    let per = (thread_count / 2).max(1);
    let producer_count = per;
    let consumer_count = per;

    let hive = build_hive::<ChannelWorker>(thread_count);
    println!("threads: {}", thread_count);
    println!("producers: {}", producer_count);
    println!("consumers: {}", consumer_count);

    let result = do_bench(&hive, producer_count, consumer_count); // warmup
    println!("output: {}", result);

    let start = Instant::now();
    for _ in 0..ITER_COUNT {
        let _ = do_bench(&hive, producer_count, consumer_count);
    }
    let dur_us = start.elapsed().as_micros();

    let elements_per_sec = (ELEMENT_COUNT as u128) * 1_000_000 / dur_us.max(1);
    println!("runs:");
    println!("  - iteration_count: {}", ITER_COUNT);
    println!("    elements: {}", ELEMENT_COUNT);
    println!("    duration: {} us", dur_us);
    println!("    elements/sec: {}", elements_per_sec);
    println!("    max_rss: {} KiB", peak_memory_usage());
}
